compiler/rustc_mir_transform/src/sink_const_assignments.rs

```diff
@@ -0,0 +1,159 @@
+use crate::MirPass;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_middle::mir::visit::{NonUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use std::collections::hash_map::Entry;
+
+pub struct SinkConstAssignments;
+
+impl<'tcx> MirPass<'tcx> for SinkConstAssignments {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let name =
+            rustc_middle::ty::print::with_no_trimmed_paths!(tcx.def_path_str(body.source.def_id()));
+        debug!("Optimizing {:?}", name);
+
+        let borrowed_locals = rustc_mir_dataflow::impls::borrowed_locals(body);
+        'outer: for block in 0..body.basic_blocks.len() {
+            let block = block.into();
+            let block_data = &body.basic_blocks[block];
+            let Some(terminator) = &block_data.terminator else { continue; };
+
+            let mut successors = Vec::new();
```

Comment:
Perhaps this should bail out when there are too many successors. It's probably not worth creating 100 new assign statements for a 100-arm match